packages/graphql/src/classes/Executor.ts

```diff
@@ -0,0 +1,215 @@
+    Neo4jGraphQLConstraintValidationError,
+    Neo4jGraphQLForbiddenError,
+    Neo4jGraphQLRelationshipValidationError,
+} from "./Error";
+import {
+    AUTH_FORBIDDEN_ERROR,
+    AUTH_UNAUTHENTICATED_ERROR,
+    DEBUG_EXECUTE,
+    RELATIONSHIP_REQUIREMENT_PREFIX,
+} from "../constants";
+
+const debug = Debug(DEBUG_EXECUTE);
+
+interface DriverLike {
+    session(config);
+}
+
+function isDriverLike(executionContext: any): executionContext is DriverLike {
+    return typeof executionContext.session === "function";
+}
+
+interface SessionLike {
+    beginTransaction(config);
+}
+
+function isSessionLike(executionContext: any): executionContext is SessionLike {
+    return typeof executionContext.beginTransaction === "function";
+}
+
+type SessionParam = {
+    defaultAccessMode?: SessionMode;
+    bookmarks?: string | string[];
+    database?: string;
+    impersonatedUser?: string;
+    fetchSize?: number;
+};
+
+type TransactionConfig = {
+    metadata: {
+        app: string;
+        // Possible values from https://neo4j.com/docs/operations-manual/current/monitoring/logging/#attach-metadata-tx (will only be user-transpiled for @neo4j/graphql)
+        type: "system" | "user-direct" | "user-action" | "user-transpiled";
+    };
+};
+
+export type ExecutionContext = Driver | Session | Transaction;
+
+export type ExecutorConstructorParam = {
+    executionContext: ExecutionContext;
+    auth: AuthContext;
+    queryOptions?: CypherQueryOptions;
+    database?: string;
+    bookmarks?: string | string[];
+};
+
+export class Executor {
+    private executionContext: Driver | Session | Transaction;
+
+    public lastBookmark: string | null;
+
+    private queryOptions: CypherQueryOptions | undefined;
+    private auth: AuthContext;
+
+    private database: string | undefined;
+    private bookmarks: string | string[] | undefined;
+
+    constructor({ executionContext, auth, queryOptions, database, bookmarks }: ExecutorConstructorParam) {
+        this.executionContext = executionContext;
+        this.lastBookmark = null;
+
+        this.queryOptions = queryOptions;
+        this.auth = auth;
+        this.database = database;
+        this.bookmarks = bookmarks;
+    }
+
+    public async execute(query: string, parameters: any, defaultAccessMode: SessionMode): Promise<QueryResult> {
+        try {
+            if (isDriverLike(this.executionContext)) {
+                const session = this.executionContext.session(this.getSessionParam(defaultAccessMode));
+                const result = await this.sessionRun(query, parameters, defaultAccessMode, session);
+                await session.close();
+                return result;
+            }
+
+            if (isSessionLike(this.executionContext)) {
+                return await this.sessionRun(query, parameters, defaultAccessMode, this.executionContext);
+            }
+
+            return await this.transactionRun(query, parameters, this.executionContext);
```

Comment:
question: what about calling `transaction.commit()` after `transaction.run()`? Is that still needed or is that a thing of the past?
(this maybe shows more that I'm not quite sure how it all works but I thought of bringing it up anyway)

Reply:
So this is a case of when a user has passed us their own transaction, which they want to be in control of - so they will have to perform the commit after the execution is complete.